src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp

```diff
@@ -0,0 +1,157 @@
+
+ private:
+  // Preprocessor logic to avoid declaring variables() functions for
+  // tags other than the three the wrapper adds (i.e., other than
+  // gr::Tags::SpacetimeMetric, GeneralizedHarmonic::Tags::Pi, and
+  // GeneralizedHarmonic:Tags::Phi)
+  using TagShift = gr::Tags::Shift<volume_dim, Frame::Inertial, DataVector>;
+  using TagSpatialMetric =
+      gr::Tags::SpatialMetric<volume_dim, Frame::Inertial, DataVector>;
+  using TagInverseSpatialMetric =
+      gr::Tags::InverseSpatialMetric<volume_dim, Frame::Inertial, DataVector>;
+  using TagExCurvature =
+      gr::Tags::ExtrinsicCurvature<volume_dim, Frame::Inertial, DataVector>;
+
+#define FUNC_DECL(r, data, elem)                                            \
+  tuples::TaggedTuple<elem> variables(                                      \
+      const tnsr::I<DataVector, volume_dim>& /*x*/, double /*t*/,           \
+      tmpl::list<elem> /*meta*/, const IntermediateVars& intermediate_vars) \
+      const noexcept;
+
+#define MY_LIST                                                              \
+  BOOST_PP_TUPLE_TO_LIST(                                                    \
+      11,                                                                    \
+      (gr::Tags::Lapse<DataVector>, TimeDerivLapse, DerivLapse, TagShift,    \
+       TimeDerivShift, DerivShift, TagSpatialMetric, TimeDerivSpatialMetric, \
+       DerivSpatialMetric, TagInverseSpatialMetric, TagExCurvature,          \
+       gr::Tags::SqrtDetSpatialMetric<DataVector>))
+
+  BOOST_PP_LIST_FOR_EACH(FUNC_DECL, _, MY_LIST)
+#undef MY_LIST
+#undef FUNC_DECL
+
+  tuples::TaggedTuple<
+      gr::Tags::SpacetimeMetric<volume_dim, Frame::Inertial, DataVector>>
+  variables(const tnsr::I<DataVector, volume_dim>& /*x*/, double /*t*/,
+            tmpl::list<gr::Tags::SpacetimeMetric<volume_dim, Frame::Inertial,
+                                                 DataVector>> /*meta*/,
+            const IntermediateVars& intermediate_vars) const noexcept;
+
+  tuples::TaggedTuple<
+      GeneralizedHarmonic::Tags::Pi<volume_dim, Frame::Inertial>>
+  variables(const tnsr::I<DataVector, volume_dim>& /*x*/, double /*t*/,
+            tmpl::list<GeneralizedHarmonic::Tags::Pi<volume_dim,
+                                                     Frame::Inertial>> /*meta*/,
+            const IntermediateVars& intermediate_vars) const noexcept;
+
+  tuples::TaggedTuple<
+      GeneralizedHarmonic::Tags::Phi<volume_dim, Frame::Inertial>>
+  variables(
+      const tnsr::I<DataVector, volume_dim>& /*x*/, double /*t*/,
+      tmpl::list<
+          GeneralizedHarmonic::Tags::Phi<volume_dim, Frame::Inertial>> /*meta*/,
+      const IntermediateVars& intermediate_vars) const noexcept;
+};
+
+template <typename SolutionType>
+inline constexpr bool operator==(const WrappedGr<SolutionType>& lhs,
+                                 const WrappedGr<SolutionType>& rhs) noexcept {
+  return dynamic_cast<const SolutionType&>(lhs) ==
```

Comment:
Either is generally fine in this case. My reason for preferring a `dynamic_cast` is that with it at least if there is a bug introduce in the code you will get a runtime error from the cast rather than UB. Since we can reasonably make the guarantee that `WrapGh<B>` has `B` as a base class a static cast would be fine. Here it's a matter of personal preference and not having to worry about the performance overhead that the runtime check in a dynamic cast adds since we shouldn't be comparing analytic solutions for equality all that often.